OrbitGl/AsyncTrack.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "AsyncTrack.h"
+
+#include "App.h"
+#include "FunctionUtils.h"
+#include "GlCanvas.h"
+#include "ManualInstrumentationManager.h"
+#include "TimeGraph.h"
+#include "capture_data.pb.h"
+
+using orbit_client_protos::FunctionInfo;
+using orbit_client_protos::TimerInfo;
+
+AsyncTrack::AsyncTrack(TimeGraph* time_graph, const std::string& name) : TimerTrack(time_graph) {
+  SetName(name);
+  SetLabel(name);
+}
+
+[[nodiscard]] std::string AsyncTrack::GetBoxTooltip(PickingId id) const {
+  const TextBox* text_box = time_graph_->GetBatcher().GetTextBox(id);
+  if (text_box == nullptr) return "";
+  ManualInstrumentationManager* manager = time_graph_->GetManualInstrumentationManager();
+  orbit_api::Event event = manager->ApiEventFromTimerInfo(text_box->GetTimerInfo());
+
+  const FunctionInfo* func =
+      GOrbitApp->GetCaptureData().GetSelectedFunction(text_box->GetTimerInfo().function_address());
+
+  std::string function_name = manager->GetString(event.id);
+
+  return absl::StrFormat(
+      "<b>%s</b><br/>"
+      "<i>Timing measured through manual instrumentation</i>"
+      "<br/><br/>"
+      "<b>Module:</b> %s<br/>"
+      "<b>Time:</b> %s",
+      function_name, FunctionUtils::GetLoadedModuleName(*func),
+      GetPrettyTime(
+          TicksToDuration(text_box->GetTimerInfo().start(), text_box->GetTimerInfo().end())));
+}
+
+void AsyncTrack::UpdateBoxHeight() {
+  box_height_ = time_graph_->GetLayout().GetTextBoxHeight();
+  if (collapse_toggle_->IsCollapsed() && depth_ > 0) {
+    box_height_ /= static_cast<float>(depth_);
+  }
+}
+
+void AsyncTrack::OnTimer(const orbit_client_protos::TimerInfo& timer_info) {
+  // Find the first row that that can receive the new timeslice with no overlap.
+  // If none of the existing rows works, add a new row.
+  uint32_t depth = 0;
+  while (max_span_time_by_depth_[depth] >= timer_info.start()) ++depth;
```

Comment:
No it doesn't assume anything. It keeps track of the current state of every row so that we know if adding a timer on that row would overlap or not. If we can't find a row with no overlap, then we create a new one.

Reply:
Looking at the code, there is a bug however, ">=" should be ">". Adding a timer at time 0 would cause an infinite loop otherwise.

Reply:
I think I see what you mean by assuming ordering. Currently, timespans are added in order of "stop" time. If the events were received in random order, then we would probably create a lot of rows. The algorithm here doesn't try to find available "gaps" in existing rows, that could be an improvement for another PR.

Reply:
Oh no ok, now I understand what you mean... If we receive a stop event before it's corresponding start event, right? I think the assumption of ordered events is made elsewhere in the code too, didn't we have buffering in place to ensure ordering? Are you talking about the fact that because we are dealing with per-core buffers events could possibly be sent out of order?

Reply:
The code can be made more robust to handle the case where an "end" event is received before its corresponding "start". Will be addressed in subsequent PR.

Reply:
> Currently, timespans are added in order of "stop" time. If the events were received in random order, then we would probably create a lot of rows. The algorithm here doesn't try to find available "gaps" in existing rows, that could be an improvement for another PR.

This is what I meant. If the events are not received in order, the number of rows explodes. It's fine to assume that events come in order if they do (and they should as the service orders uprobes events), it would be a waste to make the algorithm much more complicated for a case that doesn't happen. But we need to be aware of the reason for the assumption. Maybe add a comment about the assumption.